### DISCUSSION TITLE
Different major/minor return with block_bio_complete

### DISCUSSION THREAD

**Discussion by abbbi** [bpftrace/bpftrace#2019] [Q&A]
hi
bpftrace v0.11.3 on debian11
using the following snipped reports the right major/minor number for the device:

```
#!/usr/bin/bpftrace
tracepoint:block:block_rq_complete
{
    $minor = (args->dev >> 20);
    $major = (args->dev & ((1 << 20) - 1));
    printf("minor: %d major %d\n", $minor, $major);
}
minor: 0 major 0
minor: 8 major 0
minor: 8 major 0
minor: 8 major 0
```
using block_bio_complete however allways returns the major/minor device number for dm-0:

```
#!/usr/bin/bpftrace
tracepoint:block:block_bio_complete
{
    $minor = (args->dev >> 20);
    $major = (args->dev & ((1 << 20) - 1));
    printf("minor: %d major %d\n", $minor, $major);
}
Attaching 1 probe...
minor: 254 major 0
minor: 254 major 0
minor: 254 major 0
```

whats the reason for this?

Looking at the format, it seems to be the same:

```
root@cefix:~# grep fmt /sys/kernel/debug/tracing/events/block/block_rq_complete/format 
print fmt: "%d,%d %s (%s) %llu + %u [%d]", ((unsigned int) ((REC->dev) >> 20)), ((unsigned int) ((REC->dev) & ((1U << 20) - 1))), REC->rwbs, __get_str(cmd), (unsigned long long)REC->sector, REC->nr_sector, REC->error
root@cefix:~# grep fmt /sys/kernel/debug/tracing/events/block/block_bio_complete/format 
print fmt: "%d,%d %s %llu + %u [%d]", ((unsigned int) ((REC->dev) >> 20)), ((unsigned int) ((REC->dev) & ((1U << 20) - 1))), REC->rwbs, (unsigned long long)REC->sector, REC->nr_sector, REC->error
```

**Comment by fbs** [comment]
Can you try with the latest binary from the docker image, just to make sure its not an old bug

**Reply by abbbi** [reply]
The same:

```
root@cefix:/# ./trace 
Attaching 1 probe...
minor: 254 major 0
minor: 254 major 0
minor: 254 major 0
minor: 254 major 0
minor: 254 major 0
root@cefix:/# bpftrace --version
bpftrace v0.13.0

```

**Comment by abbbi** [comment]
Looking at the kernel sources, block_bio_complete sets another device information than
for example block_rq_complete:

block_bio_complete:
https://github.com/torvalds/linux/blob/master/include/trace/events/block.h#L248

block_rq_complete:
https://github.com/torvalds/linux/blob/master/include/trace/events/block.h#L131

but i dont know wether if this is intended?

**Reply by danobi** [reply]
I think a bio is a "high" level part of the block layer. ie. what other subsystems will use. And the block_rq_complete is the lower level part -- a driver will talk to actual hardware. That would be block layer internal. I think block devices are stackable or something too (not an expert) so I wouldn't be surprised if the major/minor numbers change. That would be my guess.
